### PR TITLE
add support for openstack identity API v3

### PR DESCRIPTION
**Pull request by daihaojun**
this pull request is for extending existing  authentication to support openstack identity API v3.

### CODE REVIEW

File: auth_v3.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"strings"
+)
+
+const (
+	AUTH_METHOD_TOKEN         = "token"
```
Comment (ncw) [inline auth_v3.go]: The aren't very go-like these constants....  Do they need to be public?  if not can you rename them `authMethodToken` etc

File: swift.go [hunk 2]
```diff
@@ -239,10 +242,10 @@ func (c *Connection) setDefaults() {
 		c.Retries = DefaultRetries
 	}
 	if c.ConnectTimeout == 0 {
-		c.ConnectTimeout = 10 * time.Second
+		c.ConnectTimeout = 60 * time.Second
```
Comment (ncw) [inline swift.go]: I think the changes from here down in this file are probably just testing changes?  If you want to propose we change the timeouts can you do it in a separate pull request please?

File: swift_test.go [hunk 3]
```diff
@@ -16,18 +16,19 @@ import (
 	"archive/tar"
 	"bytes"
 	"crypto/md5"
+	"crypto/tls"
 	"encoding/json"
 	"encoding/xml"
 	"fmt"
+	"github.com/ncw/swift"
```
Comment (ncw) [inline swift_test.go]: goimports puts these lines where they were at the end - I suggest you leave them there otherwise goimports will just put them back!

File: swift_test.go [hunk 4]
```diff
@@ -68,8 +70,14 @@ func TestTransport(t *testing.T) {
 		ApiKey = "swifttest"
 		AuthUrl = srv.AuthURL
 	}
-	tr := &someTransport{Transport: http.Transport{MaxIdleConnsPerHost: 2048}}
+	tr := &someTransport{
```
Comment (ncw) [inline swift_test.go]: This looks like testing code?  We certainly don't want `InsecureSkipVerify` on by default

File: swift_test.go [hunk 5]
```diff
@@ -122,6 +133,142 @@ func TestAuthenticate(t *testing.T) {
 	}
 }
 
+func TestV3AuthenticateWithDomainNameAndTenantId(t *testing.T) {
```
Comment (ncw) [inline swift_test.go]: Can you factor out the repeated code in the tests please?

**Comment by ncw** [comment]
Thanks for a great contribution!  I've made a few comments against lines of the code.  Can you fix it up please and rebase your branch then I'll merge it.

Thanks

Nick

File: auth_v3.go (hunk 1 above)
Comment (daihaojun) [inline auth_v3.go]: got you , will refine the code

File: swift.go (hunk 2 above)
Comment (daihaojun) [inline swift.go]: sure, as our swift server is far away from my working pc, so need to use a larger timeout value during development , i think we do not need this code merged

File: swift_test.go (hunk 4 above)
Comment (daihaojun) [inline swift_test.go]: our swift server is SSL enabled, i will add a new option for connection object

File: swift_test.go (hunk 5 above)
Comment (daihaojun) [inline swift_test.go]: got you

**Comment by daihaojun** [comment]
@ncw, thanks for review the code! I've refined the code and committed the changed code , please help to review

**Comment by ncw** [comment]
Looks great thank you :-)

I'll run it through my own tests then merge it shortly!

**Comment by ncw** [comment]
I've tested the code and it works against my v1 & v2 test servers which is a good start!

However I haven't been able to try the v3 code.  I wanted to against rackspace which support v3 but I can't get it to work (Auth errors).

Which environment variables should I set for rackspace do you think?

**Comment by daihaojun** [comment]
@ncw here is the environment I've being used to run test against our environment 

export SWIFT_API_USER=username
export SWIFT_API_KEY=password
export SWIFT_AUTH_URL=https://servername/v3/auth
export SWIFT_AUTH_INSECURE=1
export SWIFT_CONNECTION_CHANNEL_TIMEOUT=60
export SWIFT_DATA_CHANNEL_TIMEOUT=600
export SWIFT_TENANT='hosp:hpit:w-cld-dockerhub-ent-prd'
export SWIFT_TENANT_ID=f1d193757c274dca98b9cf7a71b331ad
export SWIFT_API_DOMAIN=domainname
export SWIFT_API_DOMAIN_ID=4710ddb73d9948729d17872b1ba5c884

**Comment by daihaojun** [comment]
@ncw I am trying to get a rackspace account to test the code

**Comment by ncw** [comment]
I've been looking through the rackspace docs and they don't mention v3 so maybe it isn't supported yet for cloud files.

**Comment by ncw** [comment]
I've merged this - thank you very much for the contribution.

I haven't managed to test it against a v3 server, but I don't want to hold the merge up any longer.  If any problems are found then we'll fix them as they come up.

Thanks again

Nick

**Comment by daihaojun** [comment]
@ncw many thanks!
